src/main.py

```diff
@@ -103,9 +104,20 @@ def write_candidate(perm: Permuter, result: CandidateResult) -> None:
         except FileExistsError:
             pass
     source = result.source
+
+
     assert source is not None, "Permuter._need_to_send_source is wrong!"
     with open(os.path.join(output_dir, "source.c"), "x", encoding="utf-8") as f:
-        f.write(source)
+        if no_context_output:
+            fn_str_index = source.find(perm.fn_name)
+            for i in range(fn_str_index-30, fn_str_index-1):
+                if (source[i] == '\n' or source[i] == '\r'):
+                    fn_str_index = i
+                    break
```

Comment:
extract this logic into a separate function, and use `rfind("\n")` to find the newline backward